File: app/src/main/java/com/smart/album/views/SlideImageView.java
package com.smart.album.views;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.util.Log;
import android.view.animation.Animation;
import android.view.animation.TranslateAnimation;
import androidx.appcompat.widget.AppCompatImageView;

public class SlideImageView extends AppCompatImageView {

    private OnPanningEndListener panningEndListener;

    public SlideImageView(Context context) {
        super(context);
    }

    public SlideImageView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public SlideImageView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        startPanningIfNecessary();
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
        startPanningIfNecessary();
    }

    @Override
    protected boolean setFrame(int l, int t, int r, int b) {
        boolean result = super.setFrame(l, t, r, b);
        startPanningIfNecessary();
        return result;
    }

    private void startPanningIfNecessary() {
        if (getDrawable() != null && getWidth() > 0 && getHeight() > 0) {
            Drawable drawable = getDrawable();
            int drawableWidth = drawable.getIntrinsicWidth();
            int drawableHeight = drawable.getIntrinsicHeight();

            Log.d("pan===","width=="+getWidth());
            Log.d("pan===","height=="+getHeight());
            Log.d("pan===","drawableWidth=="+drawableWidth);
            Log.d("pan===","drawableHeight=="+drawableHeight);
            if (drawableWidth > getWidth() || drawableHeight > getHeight()) {
                // 图片尺寸超过了屏幕大小
                panImage(drawableWidth, drawableHeight);
            }
        }
    }

    private void panImage(int drawableWidth, int drawableHeight) {
        float fromXDelta = 0.0f;
        float toXDelta = 0.0f;
        float fromYDelta = 0.0f;
        float toYDelta = 0.0f;

        if (drawableWidth > getWidth()) {
            toXDelta = -getWidth();
        }

        if (drawableHeight > getHeight()) {
            toYDelta = -(drawableHeight -getHeight());
        }

        TranslateAnimation animation = new TranslateAnimation(
                fromXDelta, toXDelta,
                fromYDelta, toYDelta
        );


        Log.d("pan===","fromXDelta=="+fromXDelta);
        Log.d("pan===","toXDelta=="+toXDelta);
        Log.d("pan===","fromYDelta=="+fromYDelta);
        Log.d("pan===","toYDelta=="+toYDelta);
        animation.setDuration(15000); // 动画持续时间
        animation.setFillAfter(true);
        animation.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {}

            @Override
            public void onAnimationEnd(Animation animation) {
                if (panningEndListener != null) {
                    panningEndListener.onPanningEnd();
                }
            }

            @Override
            public void onAnimationRepeat(Animation animation) {}
        });
        startAnimation(animation);
    }

    public void setOnPanningEndListener(OnPanningEndListener listener) {
        this.panningEndListener = listener;
    }

    public interface OnPanningEndListener {
        void onPanningEnd();
    }
}

File: app/src/main/java/com/smart/album/views/SlideImageView.kt
package com.smart.album.views

import android.content.Context
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.util.Log
import android.view.animation.Animation
import android.view.animation.TranslateAnimation
import androidx.appcompat.widget.AppCompatImageView

class SlideImageView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : AppCompatImageView(context, attrs, defStyleAttr) {

    private var panningEndListener: OnPanningEndListener? = null

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        startPanningIfNecessary()
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        startPanningIfNecessary()
    }

    override fun setFrame(l: Int, t: Int, r: Int, b: Int): Boolean {
        val result = super.setFrame(l, t, r, b)
        startPanningIfNecessary()
        return result
    }

    private fun startPanningIfNecessary() {
        if (drawable != null && width > 0 && height > 0) {
            val drawableWidth = drawable.intrinsicWidth
            val drawableHeight = drawable.intrinsicHeight

            Log.d("pan===", "width==$width")
            Log.d("pan===", "height==$height")
            Log.d("pan===", "drawableWidth==$drawableWidth")
            Log.d("pan===", "drawableHeight==$drawableHeight")
            if (drawableWidth > width || drawableHeight > height) {
                // 图片尺寸超过了屏幕大小
                panImage(drawableWidth, drawableHeight)
            }
        }
    }

    private fun panImage(drawableWidth: Int, drawableHeight: Int) {
        var fromXDelta = 0f
        var toXDelta = 0f
        var fromYDelta = 0f
        var toYDelta = 0f

        if (drawableWidth > width) {
            toXDelta = -width.toFloat()
        }

        if (drawableHeight > height) {
            toYDelta = -(drawableHeight - height).toFloat()
        }

        val animation = TranslateAnimation(
            fromXDelta, toXDelta,
            fromYDelta, toYDelta
        )

        Log.d("pan===", "fromXDelta==$fromXDelta")
        Log.d("pan===", "toXDelta==$toXDelta")
        Log.d("pan===", "fromYDelta==$fromYDelta")
        Log.d("pan===", "toYDelta==$toYDelta")
        animation.duration = 15000 // 动画持续时间
        animation.fillAfter = true
        animation.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {}

            override fun onAnimationEnd(animation: Animation) {
                panningEndListener?.onPanningEnd()
            }

            override fun onAnimationRepeat(animation: Animation) {}
        })
        startAnimation(animation)
    }

    fun setOnPanningEndListener(listener: OnPanningEndListener) {
        this.panningEndListener = listener
    }

    interface OnPanningEndListener {
        fun onPanningEnd()
    }
}
